_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +87,143 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
     return true;
 }
 
+class FrameStorage
+{
+private:
+    std::unordered_map<UMC::FrameMemID, UMC_VP9_DECODER::VP9DecoderFrame> m_framesStorage;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void LockResources(const UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        // lock curent frame for decode
+        VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(frame.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+        // lock frame for copy
+        if (frame.show_existing_frame)
+        {
+            VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(frame.ref_frame_map[frame.frame_to_show]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            return;
+        }
+        // lock all references
+        for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+        {
+            const UMC::FrameMemID mid = frame.ref_frame_map[ref_index];
+            if (mid >= 0)
+            {
+                VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(mid) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            }
+        }
+    }
+
+    void UnLockResources(const UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        // unlock curent frame for decode
+        VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(frame.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+        // unlock frame for copy
+        if (frame.show_existing_frame)
+        {
+            VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(frame.ref_frame_map[frame.frame_to_show]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            return;
+        }
+        // unlock all references
+        for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+        {
+            const UMC::FrameMemID mid = frame.ref_frame_map[ref_index];
+            if (mid >= 0)
+            {
+                VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(mid) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            }
+        }
+    }
+
+    void UpdateRefFrames(UMC_VP9_DECODER::VP9DecoderFrame & frame) const
```

Comment:
UpdateRefFrames should not be the method of FrameStorage class. FrameStorage tracks the list of frames in decoding state and locks all resources required for the frames. UpdateRefFrames tracks the list of current reference frames and updates the current list for the next frame.